add block_reader interface

ryan-williams opened the pull request:
[Apache Beam is interested in using fastavro instead of apache/avro](https://issues.apache.org/jira/browse/BEAM-2810), but needs an interface like I've added here, for seeking to and reading individual record-blocks in a file.

[I've published this as `fastavro-blocks:0.19.5.4` to test.pypi.org](https://test.pypi.org/project/fastavro-blocks/0.19.5.4/) (cf. [`blocks-test-pypi` branch](https://github.com/ryan-williams/fastavro/tree/blocks-test-pypi), which has [a commit on top of this PR changing the name and version](https://github.com/ryan-williams/fastavro/commit/e602bddefea6e3a369ddbd5c3695cf3d700da586)), and [used that in a Beam branch here](https://github.com/ryan-williams/beam/tree/fastavro).

Presumably it can't go into Beam until it's in real pypi, so that will have to wait until the next release here, though those seem to happen quite frequently 🎉😀

In this change I introduced a block-iteration interface, and had the standard record-iterator wrap it. I also did a few random tweaks to make _read{.pyx,_py.py} diff more cleanly against one another.

I'm new to Cython and haven't figured out when `_read_py.py` would be used instead of `_read.pyx`; a previous version of this PR failed Travis' pypy builds due to a bug I introduced in `_read_py.py`, so I guess it's covered?

scottbelden commented:
Thanks for the PR. I'll try to take a look soon and give a full review. One note is that we will want a test or two using the new functionality, but I can help with that if needed.

ryan-williams commented:
Thanks, I'm adding some tests and fixing bugs found along the way; I'll bump this when that's done

ryan-williams commented:
OK I think this is in a better state; I added a test, and cleaned up the `Block.{offset,size}` semantics so that they refer to actual file- or memory-positions and distances, whether the data is compressed or not.

ryan-williams commented:
Thinking I'll expand the tests to verify reading the records from a temp-file (currently they just read from an in-memory buffer)

scottbelden commented on fastavro/_read_py.py:
```diff
@@ -478,34 +479,68 @@ def skip_sync(fo, sync_marker):
+        return len(self.data)
+
+
+class FileObjectReader(ReaderBase):
```
I know the cython side needed a special `Reader` object in the past because the underlying cython type couldn't be used in the same way as the python type. Is the need here on the python side basically just for being able to call `len()` (as it seems the `MemoryIO` object doesn't have a way to do that)?

ryan-williams replied:
yea, I didn't have a well-thought-through story here, thanks for calling out!

in the last two commits in the latest batch (https://github.com/tebeka/fastavro/pull/208/commits/2cfe013495aa41ea02feef50e3254c55067ceb18 and https://github.com/tebeka/fastavro/pull/208/commits/ca99dcb011803ed0627d03d86d9b08c9fb16dfdf), I took a stricter eye to which of the abstractions {`ReaderBase`, `MemoryReader`, `FileObjectReader`} were needed in each of {`.pyx`, `.py`}.

afaict, none of them are necessary!

The stream-like things we deal with are:
- "files" (which can be actual file-descriptors, or in-memory buffers simulating files in tests), and
- memory-buffers representing avro blocks.

The former need `read` and `tell` methods (`tell` is used for inferring block size after reading it), and the latter need only `read`. File-descriptors and `MemoryIO`s both support both.

I'm trying to run some of the benchmarks now to see whether there's a noticeable difference between the versions with or without these abstractions.

scottbelden replied:
Great. That should reduce some complexity. We probably should have added a test if the original cython ones were needed, but it's possible it was done for some micro-optimizations. However, I think the reduced complexity and more similar to the python implementation are more beneficial.

scottbelden commented on fastavro/_read_py.py:
```diff
@@ -478,34 +479,68 @@ def skip_sync(fo, sync_marker):
+    def __init__(self):
+        self._position = 0
+
+    def position(self):
```
Should this be changed to `tell` to make it more similar to a normal `io` object?

scottbelden commented on fastavro/_read_py.py:
```diff
@@ -478,34 +479,68 @@ def skip_sync(fo, sync_marker):
 
 def deflate_read_block(fo):
     """Read block in "deflate" codec."""
-    data = read_bytes(fo, None)
```
Thanks! Been meaning to fix this

scottbelden commented on fastavro/_read_py.py:
```diff
@@ -523,50 +558,63 @@ def acquaint_schema(schema):
-        self.fo = fo
         try:
-            self._header = read_data(fo, HEADER_SCHEMA)
+            self._fo_reader = FileObjectReader(fo)
```
Can we move this to before the `try`? Just to make it clear that the `try` is only for the `read_data` call?

scottbelden commented on fastavro/_read_py.py:
```diff
@@ -588,21 +636,84 @@ def __init__(self, fo, reader_schema=None):
+            avro_reader = block_reader(fo)
+            schema = avro_reader.schema
+            for block in avro_reader:
+                process_block(record)
```
I think this should be `process_block(block)`, right?

scottbelden commented on fastavro/_read_py.py:
```diff
@@ -32,7 +32,7 @@
 
 MASK = 0xFF
-AVRO_TYPES = set([
+AVRO_TYPES = {
```
Thanks for the cleanup here

ryan-williams commented on fastavro/_read_py.py:
```diff
@@ -478,34 +479,68 @@ def skip_sync(fo, sync_marker):
 }
 
+
+def snappy_read_block(fo):
```
I believe it was to minimize spurious diffs between `_read.pyx` and `_read_py.py`. 

as a sanity check I look at them in `meld` and eyeball the differences to make sure I didn't miss anything!

ryan-williams commented:
Thanks for all the feedback! I think I addressed everything, lmk what you think!

scottbelden commented:
Thanks! This looks great. Merging this and will get a version up on pypi shortly.

swordspoet commented:
Nice job, but how can I directly get the length of each block?

ryan-williams commented:
@swordspoet IIRC, [`Block.bytes_`](https://github.com/fastavro/fastavro/blob/0.21.4/fastavro/_read_py.py#L566) is the uncompressed bytes, and [`Block.size`](https://github.com/fastavro/fastavro/blob/0.21.4/fastavro/_read_py.py#L572) is the compressed size, does that get you what you need?

(seems like they could use some docstrings…)

swordspoet commented:
@ryan-williams thank you, what I want is how many records in each block, and I want to improve speed by directly getting the numbers of record in each block. I have tried your method and got the size of block.

swordspoet commented:
@ryan-williams thank you, I found thishttps://github.com/fastavro/fastavro/blob/0.21.4/fastavro/_read_py.py#L567

scottbelden commented:
@swordspoet @ryan-williams I created an issue to better document those attributes (as well as others) when I find some time. Hopefully this will make it easier for others in the future.
